app/models/local_transaction_edition.rb

```diff
@@ -22,9 +22,9 @@ class LocalTransactionEdition < Edition
   GOVSPEAK_FIELDS = %i[introduction more_information need_to_know].freeze
 
   validate :valid_lgsl_code, if: -> { lgsl_code.present? }
-  validates :lgil_code, presence: { message: "Enter a LGIL code" }
   validates :lgsl_code, presence: { message: "Enter a LGSL code" }
-  validates :lgil_code, numericality: { only_integer: true, message: "LGIL code can only be a whole number between 0 and 999" }, if: -> { lgil_code.present? }
+  validates :lgil_code, presence: { message: "Enter a LGIL code" },
+                        numericality: { only_integer: true, message: "LGIL code can only be a whole number between 0 and 999" }
```

Comment:
I find that the below test passes whether I have this code or your previous one. So I'm not sure this change is actually necessary. Unless I'm missing something?

Reply:
Also looks like you've lost a condiditonal step in this rewriting. Though the author probably shouldn't have done `if: -> { lgil_code.present? }` and instead passed in `allow_nil: true` as an argument to numeracility.

Reply:
@kevindew if I revert the changes made to `app/models/local_transaction_edition.rb`:

```
git co HEAD^ app/models/local_transaction_edition.rb
```

I get a failure on this integration test (using govuk-docker):

```
$ bundle exec rails test test/integration/local_transaction_create_edit_test.rb
Run options: --seed 52406

# Running:

.....F....

Finished in 17.703264s, 0.5649 runs/s, 1.9205 assertions/s.

  1) Failure:
LocalTransactionCreateEditTest#test_creating_a_local_transaction_with_a_bad_LGIL_code_displays_an_appropriate_error [/govuk/publisher/test/integration/local_transaction_create_edit_test.rb:62]:
Expected false to be truthy.

10 runs, 34 assertions, 1 failures, 0 errors, 0 skips
Coverage report generated for Integration Tests, Unit Tests to /govuk/publisher/coverage. 1962 / 2986 LOC (65.71%) covered.
```

It fails on line 62 because the validation error doesn't appear on the page:

https://github.com/alphagov/publisher/blob/a67d96c64518b43472b2760efb933845bd3137cb/test/integration/local_transaction_create_edit_test.rb#L55-L63

Do you get a different result?

Reply:
Ah sorry I do get the same result as you, I must have made a mistake in my test. Sorry about that.

I think the situation isn't to do with ability to use validates multiple times and instead about the conditional, with some extra confusion with Mongoid type casting.

As I understand it, your test isn't passing because you're now getting past the presence validation. It is instead meaning that both the presence and the numericality errors occur in situations of a string being passed in.

Some examples:

With the change you've made

```ruby
  validates :lgil_code, presence: { message: "Enter a LGIL code" },
                        numericality: { only_integer: true, message: "LGIL code can only be a whole number between 0 and 999" }
```

produces:

```
irb(main):001:0> lte = LocalTransactionEdition.new(lgil_code: nil); lte.valid?; lte.errors[:lgil_code]
=> ["Enter a LGIL code", "LGIL code can only be a whole number between 0 and 999"]
irb(main):002:0> lte = LocalTransactionEdition.new(lgil_code: "word"); lte.valid?; lte.errors[:lgil_code]
=> ["Enter a LGIL code", "LGIL code can only be a whole number between 0 and 999"]
irb(main):003:0> lte = LocalTransactionEdition.new(lgil_code: 1.5); lte.valid?; lte.errors[:lgil_code]
=> ["LGIL code can only be a whole number between 0 and 999"]
```

With separate rules:

```ruby
  validates :lgil_code, presence: { message: "Enter a LGIL code" }
  validates :lgil_code, numericality: { only_integer: true, message: "LGIL code can only be a whole number between 0 and 999" }
```

also produces:

```
irb(main):001:0> lte = LocalTransactionEdition.new(lgil_code: nil); lte.valid?; lte.errors[:lgil_code]
=> ["Enter a LGIL code", "LGIL code can only be a whole number between 0 and 999"]
irb(main):002:0> lte = LocalTransactionEdition.new(lgil_code: "word"); lte.valid?; lte.errors[:lgil_code]
=> ["Enter a LGIL code", "LGIL code can only be a whole number between 0 and 999"]
irb(main):003:0> lte = LocalTransactionEdition.new(lgil_code: 1.5); lte.valid?; lte.errors[:lgil_code]
=> ["LGIL code can only be a whole number between 0 and 999"]
```

So with the conditional of before it was never getting past the presence validator.

```ruby
  validates :lgil_code, presence: { message: "Enter a LGIL code" }
  validates :lgil_code, numericality: { only_integer: true, message: "LGIL code can only be a whole number between 0 and 999" }, if: -> { lgil_code.present? }
```

produces:

```
=> ["Enter a LGIL code"]
irb(main):002:0> lte = LocalTransactionEdition.new(lgil_code: "word"); lte.valid?; lte.errors[:lgil_code]
=> ["Enter a LGIL code"]
irb(main):003:0> lte = LocalTransactionEdition.new(lgil_code: 1.5); lte.valid?; lte.errors[:lgil_code]
=> ["LGIL code can only be a whole number between 0 and 999"]
```
---

So I think we've not quite done the right thing here, as we just create two errors where there were one. But it's not quite clear what the route to take is. The ideal route would be to do validation before type casting but that seems a total pain, it doesn't seem right to set the attribute to a string. It might have to be that we accept that this is validated rather poorly and update the demo to pass in a fraction rather than a string.

It already seems a bit poorly validated as there's nothing to stop you doing a number < 0 or > 999 despite the message 🤦